### PR TITLE
Amp slidescroll experiment - Handling Native Snapping

### PR DESCRIPTION
**Pull request by sriramkrish85**
**In this PR**

- Scroll Support for snap-points

**Yet to come**

- Polyfill snapping

Part 4 for - #3465

### CODE REVIEW

File: extensions/amp-carousel/0.1/slidescroll.js [hunk 1]
```diff
@@ -38,6 +45,13 @@ export class AmpSlideScroll extends BaseCarousel {
     /** @private {!Array<!Element>} */
     this.slides_ = this.getRealChildren();
 
+    /** @private {number} */
+    this.noOfSlides_ = this.slides_.length;
+
+    /** @private @const {boolean} */
+    this.hasLooping_ =
```
Comment (cramforce) [inline extensions/amp-carousel/0.1/slidescroll.js]: This is already assigned to above.

File: extensions/amp-carousel/0.1/slidescroll.js [hunk 2]
```diff
@@ -130,6 +142,78 @@ export class AmpSlideScroll extends BaseCarousel {
...
+   * @private
+   */
+  scrollHandler_(event) {
+    if (this.scrollTimeout_) {
+      timer.cancel(this.scrollTimeout_);
+    }
+    const currentScrollLeft = this.slidesContainer_./*REVIEW*/scrollLeft;
+    if (currentScrollLeft != this.previousScrollLeft_ &&
+        !this.hasNativeSnapPoints_) {
+      //Handle custom scroll here.
```
Comment (cramforce) [inline extensions/amp-carousel/0.1/slidescroll.js]: nit: space after `//`. Should this be a TODO?

File: extensions/amp-carousel/0.1/slidescroll.js [hunk 3]
```diff
@@ -130,6 +142,78 @@ export class AmpSlideScroll extends BaseCarousel {
...
   /**
+   * Handles scroll on the slides container.
+   * @param {!Event} event Event object.
+   * @private
+   */
+  scrollHandler_(event) {
+    if (this.scrollTimeout_) {
+      timer.cancel(this.scrollTimeout_);
+    }
+    const currentScrollLeft = this.slidesContainer_./*REVIEW*/scrollLeft;
```
Comment (cramforce) [inline extensions/amp-carousel/0.1/slidescroll.js]: OK

File: extensions/amp-carousel/0.1/slidescroll.js [hunk 4]
```diff
@@ -130,6 +142,78 @@ export class AmpSlideScroll extends BaseCarousel {
...
+    if (currentScrollLeft != this.previousScrollLeft_ &&
+        !this.hasNativeSnapPoints_) {
+      //Handle custom scroll here.
+    }
+    this.scrollTimeout_ = timer.delay(() => {
```
Comment (cramforce) [inline extensions/amp-carousel/0.1/slidescroll.js]: Add some docs as to why this delay.

File: extensions/amp-carousel/0.1/slidescroll.js [hunk 5]
```diff
@@ -130,6 +142,78 @@ export class AmpSlideScroll extends BaseCarousel {
...
+      newIndex = (newIndex < 0) ? this.noOfSlides_ - 1 :
+          (newIndex >= this.noOfSlides_) ? 0 : newIndex;
+    } else {
+      newIndex = (newIndex < 0) ? 0 :
+          (newIndex >= this.noOfSlides_) ? this.noOfSlides_ - 1 : newIndex;
+    }
+    this.vsync_.mutatePromise(() => {
+      this.slidesContainer_.classList.add('no-scroll');
+      this.showSlide_(newIndex);
+    }).then(() => {
```
Comment (cramforce) [inline extensions/amp-carousel/0.1/slidescroll.js]: Why this callback?

**Comment by cramforce** [comment]
Can I try this on Heroku?

**Comment by sriramkrish85** [comment]
@cramforce - heroku url - https://ampskrish.herokuapp.com/test/manual/amp-slidescroll.amp.html
You need to enable the `amp-slidescroll` experiment

File: extensions/amp-carousel/0.1/slidescroll.js (hunk 1 above)
Comment (sriramkrish85) [inline extensions/amp-carousel/0.1/slidescroll.js]: Done

File: extensions/amp-carousel/0.1/slidescroll.js (hunk 2 above)
Comment (sriramkrish85) [inline extensions/amp-carousel/0.1/slidescroll.js]: Done

File: extensions/amp-carousel/0.1/slidescroll.js (hunk 3 above)
Comment (sriramkrish85) [inline extensions/amp-carousel/0.1/slidescroll.js]: Done

File: extensions/amp-carousel/0.1/slidescroll.js (hunk 4 above)
Comment (sriramkrish85) [inline extensions/amp-carousel/0.1/slidescroll.js]: Done - let me know if that looks okay.

File: extensions/amp-carousel/0.1/slidescroll.js (hunk 5 above)
Comment (sriramkrish85) [inline extensions/amp-carousel/0.1/slidescroll.js]: in order to stop sending scroll events we make the container `overflow: hidden` while we update the scroll position

so once the update is done - we would need to enable scrolling again.

Comment (cramforce) [inline extensions/amp-carousel/0.1/slidescroll.js]: I'm pretty sure that if you delete line `208` change `205` to just `mutate` it does the same thing. Or am I missing something?

Comment (sriramkrish85) [inline extensions/amp-carousel/0.1/slidescroll.js]: Oh Yes!

File: extensions/amp-carousel/0.1/slidescroll.js [hunk 6]
```diff
@@ -130,6 +139,79 @@ export class AmpSlideScroll extends BaseCarousel {
...
+   * @private
+   */
+  scrollHandler_(event) {
+    if (this.scrollTimeout_) {
+      timer.cancel(this.scrollTimeout_);
+    }
+    const currentScrollLeft = this.slidesContainer_./*OK*/scrollLeft;
+    if (currentScrollLeft != this.previousScrollLeft_ &&
+        !this.hasNativeSnapPoints_) {
+      // TODO(sriramkrish85)Handle custom scroll here.
```
Comment (cramforce) [inline extensions/amp-carousel/0.1/slidescroll.js]: Nit: space after `)` :)

File: extensions/amp-carousel/0.1/slidescroll.js [hunk 7]
```diff
@@ -130,6 +139,79 @@ export class AmpSlideScroll extends BaseCarousel {
...
+    let newIndex = this.slideIndex_ + updateValue;
+
+    if (this.hasLooping_) {
+      newIndex = (newIndex < 0) ? this.noOfSlides_ - 1 :
+          (newIndex >= this.noOfSlides_) ? 0 : newIndex;
+    } else {
+      newIndex = (newIndex < 0) ? 0 :
+          (newIndex >= this.noOfSlides_) ? this.noOfSlides_ - 1 : newIndex;
+    }
+    this.vsync_.mutate(() => {
```
Comment (cramforce) [inline extensions/amp-carousel/0.1/slidescroll.js]: Document this nested mutate

**Comment by cramforce** [comment]
LGTM

File: extensions/amp-carousel/0.1/slidescroll.js (hunk 6 above)
Comment (sriramkrish85) [inline extensions/amp-carousel/0.1/slidescroll.js]: Done!

File: extensions/amp-carousel/0.1/slidescroll.js (hunk 7 above)
Comment (sriramkrish85) [inline extensions/amp-carousel/0.1/slidescroll.js]: Done

**Pull request merged**